### PR TITLE
BUG: Fix KeyError in merge on CategoricalIndex

### PR DESCRIPTION
**Pull request by fjetter**
For categorical type indices a `KeyError` is raised when the index level is used during a merge on an index level

Example:

```
import pandas as pd
left = pd.DataFrame(
    {"left_data": [1, 2]},
   index=pd.CategoricalIndex(data=["A", "B"], categories=["A", "B"], name='index_col')
)
right = pd.DataFrame(
      {"right_data": [1.0, 2.0]},
   index=pd.CategoricalIndex(data=["A", "B"], categories=["A", "B"], name='index_col')
)
result = left.merge(right, on=['index_col'])
```

With this fix, the behavior of the test `test_merge_datetime_index(self, klass)` changed, though. IMHO, the behavior in this PR is more consistent since it is the same for all input types but I'm not sure what the actual behavior should be and I couldn't find a section in the documentation explaining this path.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
can u show. short example of what you are trying to do

**Comment by fjetter** [comment]
The following code raises a `KeyError` if the index is of categorical type
```
import pandas as pd
left = pd.DataFrame(
    {"left_data": [1, 2]},
   index=pd.CategoricalIndex(data=["A", "B"], categories=["A", "B"])
)
right = pd.DataFrame(
      {"right_data": [1.0, 2.0]},
   index=pd.CategoricalIndex(data=["A", "B"], categories=["A", "B"])
)
result = left.merge(right, on=['index_col'])
```

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -1826,3 +1834,51 @@ def test_merge_on_indexes(self, left_df, right_df, how, sort, expected):
                           how=how,
                           sort=sort)
         tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize('index',
+                         [
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: write like

```
@pytest.mark.parametrize(
    'index',
     [
```

so it is closer to the left margin, then each of the indexes can be writen more simply, also don't use the data kwarg
e.g.
```Index(['A', 'B'], name='index_col')``

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -1826,3 +1834,51 @@ def test_merge_on_indexes(self, left_df, right_df, how, sort, expected):
...
+                             RangeIndex(
+                                 start=0,
+                                 stop=2,
+                                 name='index_col'),
+                             DatetimeIndex(
+                                 data=["2018-01-01", "2018-01-02"],
+                                 name='index_col'),
+                         ])
+def test_merge_index_types(index):
+    left = DataFrame(
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: should be a 1-liner

**Review by jreback — CHANGES_REQUESTED** [review]
if you can fix up the formatting will have another look. pls add a whatsnew note as well. (reshaping bug fixes)

**Comment by fjetter** [comment]
@jreback I fixed the formatting and added a changelog entry for the bug I fixed. Since I didn't know for sure if the other change in behavior is intended I left it out for now. Can add another entry once this is settled

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @fjetter nice patch!

**Pull request merged**